Anomaly notifications

Knispel2 opened the pull request:
Issue: #140 

## What changes were proposed in this pull request?
Added code that creates notifications about objects with a high Anomaly score in Telegram and Slack

JulienPeloton commented on fink_filters/filter_anomaly_notification/filter.py:
```diff
@@ -0,0 +1,72 @@
+    filtred_id = set(df_min['objectId'].values)
+    df = df.withColumn('flag',
+                       when((df.objectId.isin(filtred_id)), lit(True))
+                       .otherwise(lit(False)))
```
`objectId` is not unique per alert. Several alerts from the same object share the same `objectId`. Hence here you incorporate potentially alerts with low `anomaly_score` because they have the same `objectId` than alerts with high `anomaly_score` (and that's why you probably had more alerts than the `threshold` each time).

The proposed skeleton avoids this.

Knispel2 replied:
I wanted to output all `objectId` alerts that had at least one alert with a high `anomaly_score`. I thought there shouldn't be many alerts with the same `objectId`, but they could convey useful information.

Knispel2 replied:
If I use the following approach: 
`pdf_anomalies = df.sort(['anomaly_score'], ascending=True).limit(threshold).toPandas()`
then many of the notifications have the same objectId. So, for example, out of 10 notifications upload `ftransfer_ztf_2023-03-24_521361`, only 4 are associated with a unique `objectId`.
To output only unique objectId, can I use the following code?
```
pdf_dup_group = df_proc.groupBy('objectId').min('anomaly_score')
pdf_dup_group = pdf_dup_group.sort(['min(anomaly_score)'], ascending=True).limit(threshold).toPandas()
upper_bound = np.max(pdf_dup_group['min(anomaly_score)'].values)
df_result = df_proc.withColumn('flag', df_proc['anomaly_score'] <= upper_bound)
pdf_anomalies = df_result.filter(df_result['flag']).toPandas()
```
I want to first group the data by `objectId `and then find the threshold by `anomaly_score`. Then the number of notifications exceeds the threshold, but they will all be high `anomaly_score`. I made this change in the last commit and tested it with `ftransfer_ztf_2023-03-24_521361` — I got 15 notifications at threshold=10

JulienPeloton replied:
Thanks for the detailed explanation -- it looks good to me! The only comment is that you need to scan twice the dataframe, but I think it is unavoidable (I could not come up with another solution).

JulienPeloton replied:
A potential trick -- is to speculate. Say, you know that there shouldn't be more than 10 times more alerts than the threshold, in which case, you can do the following:

```python
# speculate
pdf_anomalies_ext = df.sort(['anomaly_score'], ascending=True).limit(10 * threshold).toPandas()

# Get the upper bound
upper_bound = np.max(pdf_anomalies_ext.drop_duplicates(['objectId'])['anomaly_score'].values[:threshold]) 

# final results
pdf_anomalies = pdf_anomalies_ext[pdf_anomalies_ext['anomaly_score'] <= upper_bound]
```

This way, you scan only once the entire dataframe :-)

JulienPeloton replied:
the speculation simplifies a lot the DAG (the `Exchange` phase -- which brings high pressure on the cluster -- is removed):

```python
df_proc.sort(['anomaly_score'], ascending=True).limit(10 * threshold).explain()                                                                      
== Physical Plan ==
TakeOrderedAndProject(limit=100, orderBy=[anomaly_score#27 ASC NULLS FIRST], output=[candid#0L,schemavsn#1,publisher#2,objectId#3,candidate#4,prv_candidates#5,cutoutScience#6,cutoutTemplate#7,cutoutDifference#8,timestamp#9,cdsxmatch#10,DR3Name#11,Plx#12,e_Plx#13,gcvs#14,vsx#15,x3hsp#16,x4lac#17,mangrove#18,roid#19,rf_snia_vs_nonia#20,snn_snia_vs_nonia#21,snn_sn_vs_all#22,mulens#23,... 11 more fields])
+- FileScan parquet [candid#0L,schemavsn#1,publisher#2,objectId#3,candidate#4,prv_candidates#5,cutoutScience#6,cutoutTemplate#7,cutoutDifference#8,timestamp#9,cdsxmatch#10,DR3Name#11,Plx#12,e_Plx#13,gcvs#14,vsx#15,x3hsp#16,x4lac#17,mangrove#18,roid#19,rf_snia_vs_nonia#20,snn_snia_vs_nonia#21,snn_sn_vs_all#22,mulens#23,... 11 more fields] Batched: false, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[hdfs://134.158.75.222:8020/user/julien.peloton/archive/science/year=2023/month=..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<candid:bigint,schemavsn:string,publisher:string,objectId:string,candidate:struct<jd:double...



df_proc.groupBy('objectId').min('anomaly_score').sort(['min(anomaly_score)'], ascending=True).limit(threshold).explain()                             
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- TakeOrderedAndProject(limit=10, orderBy=[min(anomaly_score)#2589 ASC NULLS FIRST], output=[objectId#3,min(anomaly_score)#2589])
   +- HashAggregate(keys=[objectId#3], functions=[min(anomaly_score#27)])
      +- Exchange hashpartitioning(objectId#3, 200), ENSURE_REQUIREMENTS, [id=#2552]
         +- HashAggregate(keys=[objectId#3], functions=[partial_min(anomaly_score#27)])
            +- Project [objectId#3, anomaly_score#27]
               +- FileScan parquet [objectId#3,anomaly_score#27,month#33,day#34] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[hdfs://134.158.75.222:8020/user/julien.peloton/archive/science/year=2023/month=..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<objectId:string,anomaly_score:double>
```

Knispel2 replied:
Thank you very much!
That's a great idea, I used this code in the last commit

JulienPeloton replied:
Excellent -- it looks good to me!

JulienPeloton commented on fink_filters/filter_anomaly_notification/filter.py:
```diff
@@ -0,0 +1,72 @@
+    >>> print(sum(mask))
+    """
+    med = df.select('anomaly_score').approxQuantile('anomaly_score', [0.5], 0.25)
+    df_filtred = df.sort(['anomaly_score'], ascending=True).limit(min(df.count(), threshold))
```
No need for `min` -- by default Spark will look for `threshold`, otherwise it will stop before. Moreover this `min` forced you to compute the `df.count()` which was an extra overhead.

The proposed skeleton simplifies it.

Knispel2 replied:
Understood, thank you!

JulienPeloton commented on fink_filters/filter_anomaly_notification/filter.py:
```diff
@@ -0,0 +1,72 @@
+    send_data, slack_data = [], []
+    for _, row in df_min.iterrows():
+        gal = SkyCoord(ra=row.ra*u.degree, dec=row.dec*u.degree, frame='icrs').galactic
+        send_data.append(f'''ID: [{row.objectId}](https://fink-portal.org/{row.objectId})
```
In general, try to avoid code duplication. Define object once, and re-use them as much as possible.

Knispel2 replied:
OK, got it, thank you!

JulienPeloton commented on fink_filters/filter_anomaly_notification/filter.py:
```diff
@@ -0,0 +1,72 @@
+Real bogus: {round(row.rb, 2)}
+Anomaly score: {round(row.anomaly_score, 2)}''')
+    med = round(med[0], 2)
+    filter_utils.msg_handler(send_data, slack_data, med)
```
In the proposed skeleton, I propose to separate the call for Telegram and Slack. This way, users can use None, one of the two, or both.

JulienPeloton commented on fink_filters/filter_anomaly_notification/filter_utils.py:
```diff
@@ -0,0 +1,46 @@
+    url += tokens.tg_token
+    method = url + "/sendMessage"
+    tg_data = [f'Median anomaly score overnight: {med}'] + tg_data
+    slack_client = SlackClient(tokens.slack_token)
```
Where this token is defined and stored in practice?

JulienPeloton reviewed (commented):
Thanks @Knispel2  for the code addition! It is a good first trial.

So if I understand correctly, this functions takes all the data from a given night, and send to telegram or Slack the `threshold` most promising alerts according to their `anomaly_score`. 

As it is not a standard Fink filters to be used in real-time, your current implementation suffers several flaws (see review comments). So I propose to redesign the function to make clearer and more efficient. Here is a possible skeleton:

```python
def anomaly_notification_(
        df, threshold=10, 
        send_to_tg=False, channel_id=None, 
        send_to_slack=False, channel_name=None) -> DataFrame:
    """ Create event notifications with a high anomaly_score value
    
    Notes
    ----------
    Notifications can be sent to a Slack or Telegram channels.
    
    Parameters
    ----------
    df : Spark DataFrame 
        Mandatory columns are:
            objectId : unique identifier for this object
            lc_features: Dict of dicts of floats.
                Keys of first dict - filters (fid),
                keys of inner dicts - names of features
            rb: RealBogus quality score
            anomaly_score: Anomaly score
            timestamp : UTC time
    threshold: optional, int
        Number of notifications. Default is 10
    send_to_tg: optional, boolean
        If true, send message to Telegram. Default is False
    channel_id: str
        If `send_to_tg` is True, `channel_id` is the name of the 
        Telegram channel.
    send_to_slack: optional, boolean
        If true, send message to Slack. Default is False
    channel_id: str
        If `send_to_slack` is True, `channel_name` is the name of the 
        Slack channel.
        
    Returns
    ----------
    out: DataFrame
        Return the input Spark DataFrame with a new column `flag`
        for locating anomalous alerts.
    

    Examples
    ----------
    >>> from fink_utils.spark.utils import concat_col
    >>> from fink_science.ad_features.processor import extract_features_ad
    >>> from fink_science.anomaly_detection.processor import anomalys_score
    >>> df = spark.read.format('parquet').load('datatest')
    >>> what = ['magpsf', 'jd', 'sigmapsf', 'fid']
    >>> prefix = 'c'
    >>> what_prefix = [prefix + i for i in what]
    >>> for colname in what:
    ...    df = concat_col(df, colname, prefix=prefix)
    >>> df = df.withColumn('lc_features', extract_features_ad(*what_prefix, 'objectId'))
    >>> df = df.withColumn("anomaly_score", anomaly_score("lc_features"))
    >>> df_proc = df.select(
    ...     'objectId', 'candidate.ra',
    ...     'candidate.dec', 'candidate.rb',
    ...     'anomaly_score', 'timestamp')
    >>> df_out = anomaly_notification_(df_proc)
    >>> print(df_out.filter(df_out['flag']).count())
    """
    # Compute the median for the night 
    med = df.select('anomaly_score').approxQuantile('anomaly_score', [0.5], 0.25)
    med = round(med[0], 2)
    
    # Extract anomalous objects
    pdf_anomalies = df.sort(['anomaly_score'], ascending=True).limit(threshold).toPandas()
    upper_bound = np.max(pdf_anomalies['anomaly_score'].values)
            
    tg_data, slack_data = [], []
    for _, row in pdf_anomalies.iterrows():
        gal = SkyCoord(ra=row.ra*u.degree, dec=row.dec*u.degree, frame='icrs').galactic
        
        t1a = f'ID: [{row.objectId}](https://fink-portal.org/{row.objectId})'
        t1b = f'ID: <https://fink-portal.org/{row.objectId}|{row.objectId}>'
        t2 = f'GAL coordinates: {round(gal.l.deg, 6)},   {round(gal.b.deg, 6)}'
        t3 = f'UTC: {str(row.timestamp)[:-3]}'
        t4 = f'Real bogus: {round(row.rb, 2)}'
        t5 = f'Anomaly score: {round(row.anomaly_score, 2)}'
        tg_data.append(f'''{t1a}
{t2}
{t3}
{t4}
{t5}''')
        slack_data.append(f'''{t1b}
{t2}
{t3}
{t4}
{t5}''')
    
    if send_to_slack:
        filter_utils.msg_handler_slack(slack_data, channel_name, med)
    if send_to_tg:
        filter_utils.msg_handler_tg(tg_data, channel_id, med)
    
    df = df.withColumn('flag', df['anomaly_score'] <= upper_bound)
    return df
```

Feel free to discuss this implementation. Note that I haven't check it properly as I haven't access to the telegram token. 

For the Slack part, I'm not sure it works correctly as it? If not, I would propose that you focus on the Telegram implementation, and I will add the Slack one (with correct token) after merging.

Knispel2 commented on fink_filters/filter_anomaly_notification/filter_utils.py:
```diff
@@ -0,0 +1,46 @@
+    try:
+        channels = slack_client.api_call("conversations.list")['channels']
+        for channel in channels:
+            if channel['name'] == 'fink_alert':
```
I have created a separate workspace in Slack to test the script (link to join: https://join.slack.com/t/finkbottest/shared_invite/zt-1sj6pah1h-D6lw4hW6Jq_1lvseu5tqVQ ), and the script successfully sends messages to this channel. However, this is my first time encountering Slack, so it really may not be the best implementation

JulienPeloton replied:
It looks good to me -- you can customise it later if need be,

Knispel2 commented on fink_filters/filter_anomaly_notification/filter_utils.py:
```diff
@@ -0,0 +1,46 @@
+def msg_handler(tg_data, slack_data, med):
+    url = "https://api.telegram.org/bot"
+    channel_id = "@ZTF_anomaly_bot"
+    url += tokens.tg_token
```
The token is placed in the file tokens.py in /fink-filters/filter_anomaly_notification/, I have now emailed this file to you

Knispel2 commented on fink_filters/filter_anomaly_notification/filter.py:
```diff
@@ -0,0 +1,72 @@
+
+
+
+def anomaly_notification_(df, threshold=10) -> pd.Series:
```
Understood, thank you!

JulienPeloton commented:
Note that the CI is failing because `SlackClient` is not installed:
https://github.com/astrolabsoftware/fink-filters/actions/runs/4617903685/jobs/8197100439?pr=141
I will regenerate the Docker images to include it.

JulienPeloton commented on fink_filters/filter_anomaly_notification/filter_utils.py:
```diff
@@ -0,0 +1,49 @@
+import time
+import requests
+from slackclient import SlackClient
+import tokens
```
In production, those tokens must not be loaded from a module sitting next to the code, but rather passed as runtime CLI arguments to Spark for security reasons. I propose the following. First import `os` instead of `tokens`:

```python
import os
```

Then later in the code, replace 

```python
tokens.slack_token
tokens.tg_token
```
by 
```
os.environ['ANOMALY_SLACK_TOKEN']
os.environ['ANOMALY_TG_TOKEN']
```

Then, in practice, you would launch the Spark job using:

```bash
spark-submit --master [...]
--conf spark.executorEnv.ANOMALY_SLACK_TOKEN="<your token for Slack>"\
--conf spark.executorEnv.ANOMALY_TG_TOKEN="<your token for Telegram>"\
[...]
```
This way the tokens will not have to be shipped with the code.

Knispel2 replied:
Understood, thank you! Did it in the last commit

JulienPeloton commented on fink_filters/filter_anomaly_notification/filter.py:
```diff
@@ -0,0 +1,102 @@
+    med = df_proc.select('anomaly_score').approxQuantile('anomaly_score', [0.5], 0.05)
+    med = round(med[0], 2)
+    # Extract anomalous objects
+    trick_par = 10
```
What about defining this as an argument of `anomaly_notification_`? This way, it can be easily tuned if need be.

Knispel2 replied:
OK, did it in the last commit

JulienPeloton commented on fink_filters/filter_anomaly_notification/filter_utils.py:
```diff
@@ -0,0 +1,49 @@
+            )
+        time.sleep(3)
+
+def msg_handler_tg(tg_data, channel_id, med):
```
Could you add a docstring to this function?

Knispel2 replied:
Done!

JulienPeloton reviewed (commented):
Thanks @Knispel2 for the change -- they look all good to me! I left a final comment about the tokens (and minor comments about the doc).

Knispel2 commented on fink_filters/filter_anomaly_notification/filter_utils.py:
```diff
@@ -0,0 +1,49 @@
+import tokens
+
+
+def msg_handler_slack(slack_data, channel_name, med):
```
Done!

Knispel2 commented:
> By the way, the Slack client syntax you are using is no more valid. It used to be called `slackclient` in the v2, but they renamed the package into `slack_sdk` and changed the syntax in the current (v3) version:
> 
> https://pypi.org/project/slackclient/ https://github.com/slackapi/python-slack-sdk https://github.com/slackapi/python-slack-sdk#migrating-from-v2
> 
> Could you update the syntax for using the new version?

Sorry, I seem to have used outdated documentation. I've fixed it now!

JulienPeloton reviewed (approved):
Thanks @Knispel2 -- all good to me, I am merging! Thank you for the work.
